### ISSUE TITLE
autoProcessMedia.cfg is getting overwritten

### ISSUE THREAD

**Issue by brianjmurrell** [clinton-hall/nzbToMedia#1335]
So, I make a bunch of changes to my `autoProcessMedia.cfg` file and come back a while later only to see the changes reverted and a modified timestamp on the file at a time when I was sleeping.

Are there any conditions/situations where nzbToMedia writes this file out?

**Comment by brianjmurrell** [comment]
I guess it's `config.migrate()` in `core/nzbToMediaConfig.py`:

```
Traceback (most recent call last):
File "nzbToSickBeard.py", line 260, in <module>
result = nzbToMedia.main(sys.argv, section)
File "nzbToMedia.py", line 649, in main
core.initialize(section)
File "core/__init__.py", line 283, in initialize
if not config.migrate():
File "core/nzbToMediaConfig.py", line 246, in migrate
CFG_NEW.write()
File "libs/configobj/configobj.py", line 2128, in write
with open(self.filename, 'wb') as h:
IOError: [Errno 13] Permission denied: 'autoProcessMedia.cfg'
```

So if nzbToMedia is going to keep writing this file out, how do I make changes to it that I want to persist?

For example, this migration keeps writing out the following transcode values (not the complete set, but these are the ones it keeps adding):

```
    outputVideoPreset = medium
    outputVideoResolution = 1920:1080
    outputVideoFramerate = 24
    outputVideoBitrate = 800000
    outputVideoCRF = 19
    outputVideoLevel = 3.1
    outputAudioCodec = ac3
    AudioCodecAllow = ""
    outputAudioChannels = 6
    outputAudioBitrate = 640k
    outputAudioTrack2Codec = libfaac
    AudioCodec2Allow = ""
    outputAudioTrack2Channels = 2
    outputAudioTrack2Bitrate = 128000
    outputAudioOtherCodec = libmp3lame
    AudioOtherCodecAllow = ""
    outputAudioOtherChannels = ""
    outputAudioOtherBitrate = 128000
    outputSubtitleCodec = ""
```

when I don't want any of those.  I comment them all out and the next `config.migrate()` brings them all back.

All I want to do is transcode any file not in the `VideoCodecAllow = "libx264,h264,h.264,AVC,avc,mpeg4,msmpeg4,MPEG-4"` list using ffmpeg's default values for all of the above, preserving the resolution of the original as well as any audio and/or subtitle tracks.

**Comment by clinton-hall** [comment]
you can't comment options out. Every option must exist, or it will always revert to the default value.
instead of commenting out the option, just set the option blank
```
    outputVideoPreset = ""
    outputVideoResolution = ""
    outputVideoFramerate = ""
    outputVideoBitrate = ""
    outputVideoCRF = ""
    outputVideoLevel =""
````
also, to prevent the migrate error, make sure that the .cfg file has read and write permissions for the user that runs nzbToMedia (the user that runs the downloader)

**Comment by brianjmurrell** [comment]
Ahh.  OK.  Subtle difference, but it is meaningful, I can understand.

So in doing this, I seem to have activated _NOOP_ transcodes such as:

```
[22:22:29] [INFO]::MAIN: Checking for files to be transcoded
[22:22:29] [DEBUG]::MAIN: calling command: /usr/bin/ffprobe -v quiet -print_format json -show_format -show_streams -show_error /Download/Foo.Bar.S01E01.720p.HDTV.X264-DIMENSION/Foo.Bar.S01E01.720p.HDTV.X264-DIMENSION.mkv
[22:22:29] [INFO]::MAIN: Transcoding video: /Download/Foo.Bar.S01E01.720p.HDTV.X264-DIMENSION/Foo.Bar.S01E01.720p.HDTV.X264-DIMENSION-transcoded.mkv
[22:22:29] [DEBUG]::MAIN: calling command: nice -n0 ionice -c0 -n0 /usr/bin/ffmpeg -loglevel warning -report -fflags +genpts -i /Download/Foo.Bar.S01E01.720p.HDTV.X264-DIMENSION/Foo.Bar.S01E01.720p.HDTV.X264-DIMENSION.mkv -map 0:0 -map 0:1 -c:v copy -c:a:0 copy -sn /Download/Foo.Bar.S01E01.720p.HDTV.X264-DIMENSION/Foo.Bar.S01E01.720p.HDTV.X264-DIMENSION-transcoded.mkv
[22:23:07] [INFO]::MAIN: Transcoding of video to /Download/Foo.Bar.S01E01.720p.HDTV.X264-DIMENSION/Foo.Bar.S01E01.720p.HDTV.X264-DIMENSION-transcoded.mkv succeeded
```

As you can see, it only took 38 seconds and according to `ffprobe` on the two files, it didn't change anything.

So why did it even happen?

My _transcode_ settings are:

```
[Transcoder]
    # getsubs. enable to download subtitles.
    getSubs = 0
    # subLanguages. create a list of languages in the order you want them in your subtitles. 
    subLanguages = eng, spa, fra
    # transcode. enable to use transcoder
    transcode = 1
    ###### duplicate =1 will create a new file. =0 will replace the original
    duplicate = 1
    # concat. joins cd1 cd2 etc into a single video.
    concat = 1
    # IgnoreExtensions is a comma-separated list of extensions that will not be transcoded.
    ignoreExtensions = ""
    # outputFastStart. 1 will use -movflags + faststart. 0 will disable this from being used.
    outputFastStart = 0
    # outputQualityPercent. used as -q:a value. 0 will disable this from being used.
    outputQualityPercent = 0
    # outputVideoPath. Set path you want transcoded videos moved to. Leave blank to disable.
    outputVideoPath = ""
    # processOutput. 1 will send the outputVideoPath to SickBeard/CouchPotato. 0 will send original files.
    processOutput = 0
    # audioLanguage. set the 3 letter language code you want as your primary audio track.
    audioLanguage = eng
    # allAudioLanguages. 1 will keep all audio tracks (uses AudioCodec3) where available.
    allAudioLanguages = 0
    # allSubLanguages. 1 will keep all existing sub languages. 0 will discard those not in your list above.
    allSubLanguages = 0
    # embedSubs. 1 will embed external sub/srt subs into your video if this is supported.
    embedSubs = 1
    # burnInSubtitle. burns the default sub language into your video (needed for players that don't support subs)
    burnInSubtitle = 0
    # extractSubs. 1 will extract subs from the video file and save these as external srt files.
    extractSubs = 0
    # externalSubDir. set the directory where subs should be saved (if not the same directory as the video)
    externalSubDir = ""
    # hwAccel. 1 will set ffmpeg to enable hardware acceleration (this requires a recent ffmpeg)
    hwAccel = 0
    # generalOptions. Enter your additional ffmpeg options here with commas to separate each option/value (i.e replace spaces with commas).
    generalOptions = -report
    # outputDefault. Loads default configs for the selected device. The remaining options below are ignored.
    # If you want to use your own profile, leave this blank and set the remaining options below.
    # outputDefault profiles allowed: iPad, iPad-1080p, iPad-720p, Apple-TV2, iPod, iPhone, PS3, xbox, Roku-1080p, Roku-720p, Roku-480p, mkv, mp4-scene-release
    outputDefault = ""
    #### Define custom settings below.
    outputVideoExtension = .mkv
    outputVideoCodec = libx264
    VideoCodecAllow = "libx264,h264,h.264,AVC,avc,mpeg4,msmpeg4,MPEG-4"
    outputVideoPreset = ""
    outputVideoResolution = ""
    outputVideoFramerate = ""
    outputVideoBitrate = ""
    outputVideoCRF = ""
    outputVideoLevel = ""
    outputAudioCodec = ""
    AudioCodecAllow = ""
    outputAudioChannels = ""
    outputAudioBitrate = ""
    outputAudioTrack2Codec = ""
    AudioCodec2Allow = ""
    outputAudioTrack2Channels = ""
    outputAudioTrack2Bitrate = ""
    outputAudioOtherCodec = ""
    AudioOtherCodecAllow = ""
    outputAudioOtherChannels = ""
    outputAudioOtherBitrate = ""
    outputSubtitleCodec = ""
```

The goal being to transcode any files whose video stream is not one of the _VideoCodecAllow_ types, leaving everything else (i.e. resolution, audio streams, subtitile streams, etc.) as is.

Granted the _NOOP_ transcode is short but I'd rather not increase the risk of a good media file getting turned into a bad media file through even a _NOOP_ transcode process.

**Comment by clinton-hall** [comment]
The video was within the VideoCodecAllow, and you haven't defined a preferred audio codec, so in this case, there was nothing to be done.
I don't know if what you want can be achieved. This part script was designed to to always convert the file to a defined output.

**Comment by brianjmurrell** [comment]
Indeed, I am not arguing that the transcode it did was wrong.  If the code is **always** going to transcode then it did what it should.  I just didn't expect a transcode to be done at all if the format/parameters of the existing file meets the transcode parameters already.  Even the name of the parameter _VideoCodecAllow_ suggests to me code something like:

```python
if video_codec not in VideoCodecAllow.split(",")
    transcode_video()
```

I.e. don't transcode if the video's encoding is already one of _VideoCodecAllow_.

I never used to get even NOOP transcoding before my most recent changes.  But I guess this all changed when I also changed `ignoreExtensions = .avi, .mkv, .mp4` to `ignoreExtensions = ""` so that HEVC (typically in `.mkv`) files didn't get ignored and got transcoded to _h.264_.

Anyway, I guess ultimately, this is now an [_RFE_](https://en.wikipedia.org/wiki/Change_request) to skip transcoding if the target already meets the source requirements.  I.e. if the transcode would end up being a NOOP transcode.  :-)

Thanks much for all of your input on this.  Much appreciated.  I learned something new which is always great!
